feat(start_from_here): add execution history replay functionality

MSherbinii opened the pull request:
## Goal
  When debugging state machines with long execution paths or complex parallel branches, developers often need to test specific states repeatedly. Currently, they must restart the entire state machine
  each time. This feature allows starting execution from any historical execution point, saving significant debugging time.

  ## Changes
  - Add `replay_from_history()` method to ExecutionEngine for starting execution from history items
  - Build runtime context map from execution history to restore scoped_data at target timestamp
  - Add context menu "Start From Here" option in execution history tree view
  - Restore scoped data for container states during replay to maintain correct runtime state
  - Handle parallel branch states by collecting data from all concurrent branches at replay point

  ## How It Works
  When a user right-clicks on a CallItem in the execution history and selects "Start From Here", the system:
  1. Reconstructs the runtime state (scoped_data or scoped_variables) at that historical point by walking backwards through execution history from history_items, IN_MEMORY_EXECUTION must be enabled in config, to retrieve historical data from python object.
  2. Handles parallel execution branches by collecting state from all concurrent branches at the target timestamp
  3. Sets the execution start path and begins step-mode execution from that point

  ## Testing
  1. Run a state machine with nested hierarchy states
  2. After completion, open the execution history view
  3. Right-click on any CallItem and select "Start From Here"
  4. Verify execution begins from that point with correct scoped_data/variables and end results are as expected

JohannesErnst commented on source/rafcon/core/execution/execution_history_items.py:
```diff
@@ -263,4 +263,73 @@ def destroy(self):
+CallType = Enum('METHOD_NAME', 'EXECUTE CONTAINER')
+
+
+def build_runtime_map(target_history_item):
```
Looking at `execution_history_items.py`, I think this should rather be a member function of `HistoryItem` instead of a standalone function. Then rather calling it like `SomeHistoryItem.build_runtime_map()`.
@flolay what do you think?

flolay replied:
I personally don't have a strong opinion. Since it only uses a history item and it is associated to it.
I think making it a class function but keeping it functional and making it a static method makes sense in my eyes

JohannesErnst commented on source/rafcon/core/execution/execution_history_items.py:
```diff
@@ -263,4 +263,73 @@ def destroy(self):
+            runtime_map[item.path] = dict(item.scoped_data)
+            seen_paths.add(item.path)
+
+    if isinstance(target_history_item, ScopedDataItem):
```
Can the `target_history_item` ever be something else than a `ScopedDataItem`? Is this if needed?

MSherbinii replied:
You're correct, its not needed, initially had it before fully understanding actual RAFCON architecture.

JohannesErnst commented on source/rafcon/core/states/container_state.py:
```diff
@@ -276,6 +276,29 @@ def setup_run(self):
+    def _restore_replay_context(self):
+        """Restore scoped data from execution history replay context
+        """
+        from rafcon.core.singleton import state_machine_execution_engine
```
This is already imported for `container_state.py`

MSherbinii replied:
You're right, redundant import will remove.

JohannesErnst replied:
This still needs to be removed

JohannesErnst reviewed (commented):
Looks overall nice to me, please see few minor points

flolay commented on source/rafcon/core/execution/execution_engine.py:
```diff
@@ -285,6 +290,48 @@ def step_out(self):
         else:
             self.set_execution_mode(StateMachineExecutionStatus.FORWARD_OUT)
 
+    def set_start_path(self, state_path):
```
you introduce this function, and since it occurs now multiple times, can you please make sure that for the same behavior this function is used (e.g. at https://github.com/DLR-RM/RAFCON/blob/master/source/rafcon/core/execution/execution_engine.py#L142)

but please check if something like this strategy is used somewhere else

flolay commented on source/rafcon/core/execution/execution_history_items.py:
```diff
@@ -263,4 +263,73 @@ def destroy(self):
+        if target_parent_path and target_parent_path in ancestor_path_map:
+            runtime_map[target_parent_path] = dict(target_history_item.scoped_data)
+
+    return runtime_map
```
why do you build up such a huge runtime map again?


In the end you only need one scoped data instance (that you then update in the container state). why can you not just use this scoped data? you do know the container state from which to start right?

(this is really a genuine question, if there is a good reasoning why such a huge map is required, we can keep it but it is a ugly nested while loop which I think adds a lot of complexity.)

JohannesErnst commented:
I quickly tested the feature and it seems to work very nice! One thing that I came across so far:
- The info: `Replaying execution from QBYTMD/RYZGLE/EAWGCM/VLXMXE/MQRVHE/USFTMR` should be in human readable form so replace the IDs with the actual state names would probably be better. Otherwise this info message doesn't help the user a lot.

MSherbinii commented:
Would you prefer that we keep both pieces of information? For example:

- Shown to the user: “Replaying execution from State X” (human-readable)
- Debug output: the full state path with the IDs, for troubleshooting

JohannesErnst commented:
Yes, good idea! Maybe on "info-level" the state names and on "debug-level" the IDs 
But still show the whole path for the info-level: Like `demo1/hierarchy1/execution_state_14`

MSherbinii committed:
refactor(execution_engine): add set_start_path method to reduce code duplication
- Add set_start_path() method to ExecutionEngine to encapsulate path building logic
- Refactor start() to use set_start_path() instead of inline path building
- Refactor _prepare_run_selected_state() to use set_start_path() for consistency
- Improves code reusability and maintainability

MSherbinii committed:
feat(start_from_here): add replay context for scoped data restoration
- Add _replay_context field to ExecutionEngine to store scoped data during replay
- Add _restore_replay_context() method to ContainerState for restoring scoped data
- Call _restore_replay_context() in setup_run() to restore state when replaying
- Allows container states to maintain runtime context when starting from history

MSherbinii committed:
feat(start_from_here): add replay_from_history method for history-based execution
- Add replay_from_history() method to ExecutionEngine for starting execution from history
- Build human-readable state name path for info-level logging (e.g., demo1/hierarchy1/state)
- Log state ID path at debug level for troubleshooting
- Store scoped_data from history item in replay context for restoration
- Use set_start_path() to configure execution starting point
- Start execution in step mode from the selected history point
- Clear replay context when starting normal execution to prevent interference

MSherbinii committed:
feat(start_from_here): add Start From Here context menu to execution history
- Add "Start From Here" menu item to execution history right-click context menu
- Show menu option only for CallItem instances in the execution history
- Add _on_start_from_here() handler to process user selection
- Validate execution state before allowing replay (must be stopped/finished)
- Get active state machine from GUI model
- Call ExecutionEngine.replay_from_history() with selected history item
- Provide user feedback via logger warnings if operation cannot proceed

MSherbinii commented:
I have made more modular commits, revised the approach to apply all the comments by reviewers and follow a more digestable approach so its easier to follow changes and the decisions behind them.

JohannesErnst commented on source/rafcon/core/execution/execution_engine.py:
```diff
@@ -285,6 +280,54 @@ def step_out(self):
+            state = state.parent
+        
+        human_readable_path = "/".join(state_names)
+        logger.info("Replaying execution from {0}".format(human_readable_path))
```
This should also show the root_state not only from the first hierarchy level.

JohannesErnst commented on source/rafcon/core/execution/execution_engine.py:
```diff
@@ -285,6 +280,54 @@ def step_out(self):
+        :param state_path: Path to the state to start from (e.g., "ROOT/CONTAINER/STATE")
+        """
+        self.start_state_paths = []
+        if state_path:
```
Is this `if` really needed?
In normal operation, can this function ever be called with `state_path=None`? 
If this is not a case that can happen, I would remove the `if` otherwise this would swallow an error case.

MSherbinii replied:
This if was already in the original code(lines 140-150) . I just moved it into the new method.

`state_path=None` is a valid case - it means "start from root". For example, `start()` has `start_state_path=None` as the default parameter. When None, we want an empty list, not a crash from calling `.split()` on None.

  When state_path is not None, it's a path like "ROOT/CONTAINER/STATE" ,used in `_prepare_run_selected_state()` and `replay_from_history() ` (new function for new feature)

JohannesErnst replied:
Okay, thanks for the clarification!

JohannesErnst commented on source/rafcon/core/execution/execution_engine.py:
```diff
@@ -285,6 +280,54 @@ def step_out(self):
+        state_names = []
+        while state and not state.is_root_state:
+            state_names.insert(0, state.name)
+            state = state.parent
```
This list of `state_names` is correctly created but confusing for humans to read:
If the root state has a library (called e.g. `task 2`) which is a hierarchy state (e.g. also called `task 2`), than this list will show `root_state/task 2/task 2`. 
This is confusing because the library which is the hierarchy state one level below is mentioned twice. Of course the library can be renamed in the root state than the list will show `root_state/task 2 renamed/task 2`.
I would suggest to skip the name of the hierarchy state, if it is included as a library state.
Then, this list should show `root_state/task 2 renamed` only.

For the debug level output (IDs) I think it is okay to keep both.

JohannesErnst commented on source/rafcon/core/states/container_state.py:
```diff
@@ -276,6 +276,15 @@ def setup_run(self):
+        if state_machine_execution_engine._replay_context:
+            scoped_data = state_machine_execution_engine._replay_context.get('scoped_data')
+            if scoped_data:
+                self._scoped_data.update(scoped_data)
```
This only sets the scoped data for the current hierarchy level but upper hierarchy level scoped data is not set.

JohannesErnst reviewed (commented):
The current implementation is not working for deeper hierarchies:
Consider this state machine:
<img width="1058" height="1050" alt="latest-screenshot" src="https://github.com/user-attachments/assets/efb992b5-e365-43ab-9040-0951f6ef28c2" />

If we start the execution from inside `example set some data` the correct scoped data is set inside there. But once the execution reaches `ExecutionState 2`, the default input is used instead of the one passed from `ExecutionState 1`. 
We need to somehow recursively walk through the hierachies and set the scoped data for each hierarchy correctly.

MSherbinii committed:
fix(start_from_here): correct scoped_data restoration for nested states

- Store scoped_data separately per ancestor in runtime_map
- Skip restoration for descendants to allow fresh computation
- Remove library state duplicates from human-readable paths

JohannesErnst committed:
fix(start_from_here): minor beautification of info output

MSherbinii committed:
fix(start_from_here): improve scoped_data restoration for target state

- Start iteration from target CallItem itself instead of prev
- Build paths_to_restore from index 2 (skip root) to include target
- Remove descendant-skip logic in container_state restoration
- Clean up runtime_map entries after restoration to prevent reuse
- Ensures target state receives its scoped_data correctly

JohannesErnst committed:
fix(start_from_here): building runtime_map

Previous runtime_map was not storing the correct scoped data.
Now, runtime_map saves the most recent CallItem for each hierarchy root.
This can be used later to restore the replay context.

TODOs:
- test reliability
- right now, the execution engine pauses AFTER the selected state machine not before as expected

JohannesErnst commented:
The previous version was still not working. I went over the code and handled the `runtime_map` building. 
I tested some scenarios but we need to test more extensively.
Currently, the state machine is paused after the selected item not before as expected. This still needs to be improved.

JohannesErnst committed:
fix(start_from_here): minor bug fixes

We still need to fix that it highlights the correct state as yellow "paused".

JohannesErnst committed:
fix(start_from_here): removed start from here button for 'Enter' histroy items

MSherbinii committed:
fix(start_from_here): apply expected highlighting pattern and pause before selected state to start from directly

JohannesErnst commented on source/rafcon/core/states/container_state.py:
```diff
@@ -1099,20 +1099,44 @@ def set_start_state(self, state):
+
+                    # Find the previous state in the transition chain
+                    # This gives the yellow border to the correct state (the one before target)
+                    previous_state = self._find_previous_state_for(state)
```
I think we need to scope this function call better: right now it is also called when the user presses `run from selected state` on any state in the root state. It should only be called in the `start from execution history` case.

JohannesErnst commented on source/rafcon/core/states/container_state.py:
```diff
@@ -1099,20 +1099,44 @@ def set_start_state(self, state):
+        # Look through all transitions to find which state transitions to the target
+        for transition_id, transition in self.transitions.items():
+            # Check if this transition goes to the target state
+            if transition.to_state == target_state.state_id:
```
This doesn't necessarily select the correct state that ran before: Two or more states can all have the same transition to the target state. This right now just selects the first it finds. We can get the information about which state really ran before the selected one by looking at the execution history.

JohannesErnst reviewed (changes requested):
Thanks for the progress, I think using the `RUN_TO_SELECTED_STATE` is a good option now. However, there are remaining problems with coloring the previous ran item, please refer to comments.

MSherbinii committed:
fix(start_from_here): implement correct state highlighting and replay ctx clearing

JohannesErnst commented on source/rafcon/core/execution/execution_engine.py:
```diff
@@ -346,10 +346,26 @@ def replay_from_history(self, history_item, state_machine):
+                prev_path_parts = prev_item.path.split('/')
+                # Check if this is a sibling (same parent) by comparing path depths
+                if len(prev_path_parts) == len(target_path_parts) and \
+                   prev_path_parts[:-1] == target_path_parts[:-1]:
```
minor indentation mismatch

MSherbinii committed:
Minor indentation fix

MSherbinii committed:
test: add unit tests for start from here feature

JohannesErnst committed:
fix(tests): improved performance of start_from_here unit tests

- removed waits and info prints in test state machines
- combined multiple tests in one function to not reload state machine for each test
- removed one redundant test

JohannesErnst reviewed (approved):
Passed tests and resolved comments, ready to merge to develop branch
